cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/input/BasicPartitionedFileSetInputContext.java

```diff
@@ -42,34 +51,75 @@
 
   private static final Type STRING_PARTITION_KEY_MAP_TYPE = new TypeToken<Map<String, PartitionKey>>() { }.getType();
 
-  private final Path inputPath;
-  private final String mappingString;
+  private final Map<String, PartitionKey> pathToPartitionMapping;
 
-  private PartitionKey partitionKey;
+  private final boolean isCombineInputFormat;
+  private final Configuration conf;
+
+  private final Supplier<Path[]> inputPaths;
+  private List<PartitionKey> partitionKeys;
+
+  // for caching in case of CombineFileInputFormat
+  private String currentInputfileName;
+  private PartitionKey currentPartitionKey;
 
   BasicPartitionedFileSetInputContext(MultiInputTaggedSplit multiInputTaggedSplit) {
     super(multiInputTaggedSplit.getName());
 
     InputSplit inputSplit = multiInputTaggedSplit.getInputSplit();
-    if (!(inputSplit instanceof FileSplit)) {
-      throw new IllegalArgumentException(String.format("Expected a '%s', but got '%s'.",
-                                                       FileSplit.class.getName(), inputSplit.getClass().getName()));
+    if (inputSplit instanceof FileSplit) {
+      isCombineInputFormat = false;
+      Path path = ((FileSplit) inputSplit).getPath();
+      inputPaths = Suppliers.ofInstance(new Path[] { path });
+    } else if (inputSplit instanceof CombineFileSplit) {
+      isCombineInputFormat = true;
+      inputPaths = Suppliers.ofInstance(((CombineFileSplit) inputSplit).getPaths());
+    } else {
+      throw new IllegalArgumentException(String.format("Expected either a '%s' or a '%s', but got '%s'.",
+                                                       FileSplit.class.getName(), CombineFileSplit.class.getName(),
+                                                       inputSplit.getClass().getName()));
     }
-    this.inputPath = ((FileSplit) inputSplit).getPath();
-    this.mappingString = multiInputTaggedSplit.getConf().get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
+
+    this.conf = multiInputTaggedSplit.getConf();
+    String mappingString = conf.get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
+    this.pathToPartitionMapping =
+      GSON.fromJson(Objects.requireNonNull(mappingString), STRING_PARTITION_KEY_MAP_TYPE);
   }
 
   @Override
   public PartitionKey getInputPartitionKey() {
-    if (partitionKey == null) {
-      partitionKey = getPartitionKey(inputPath.toUri());
+    if (isCombineInputFormat) {
+      // org.apache.hadoop.mapreduce.lib.input.CombineFileRecordReader sets this in its initNextRecordReader method
+      String inputFileName = conf.get(MRJobConfig.MAP_INPUT_FILE);
+      Preconditions.checkNotNull(inputFileName);
```

Comment:
Might be better to throw with a more meaningful message and document that as a possible exception being thrown, which the user can handle (otherwise the job will fail, right?).